### PR TITLE
IOSDOX-12139 Update SDK to Swift and made into Swift Package

### PR DESCRIPTION
**Pull request by ronnie3786**
Jira Story: https://doximity.atlassian.net/browse/IOSDOX-12139

## Overview
I converted the legacy SDK from Objective-C to Swift and converted this to be a Swift Package to be used with SPM. I also kept the ability for users to do a manual integration and added the necessary Objc helper functions to keep this compatible with Objective-C.

## Technical Notes
Changes made:
- Converted the code to Swift
- Added Objective-C bridging
- Setup Swift Package details
- Removed Cocoapods and Carthage support (and mention of the README
- Added Unit Tests (updated implementation to accept a `MockApplication`
- Updated `gitignore` to ignore some build and local files

## Dev (as QA) Instructions
- Create a new Xcode Project for a basic iOS App
- Follow the SPM integration steps and usage from the updated README (for testing, you should be able to specify this branch name)
- Test the `DoxDialerCaller.shared.dialerIcon` and `DoxDialerCaller.shared.dialPhoneNumber("4254443333")` functionality
- [ ] Ensure the icon is visible as expected and the call will correctly deep-link into our Doximity app (if installed) or link to the App Store page
- Remove the SPM integration
- Follow the Manual Integration steps
- [ ] Ensure the manual integration works as expected and the same as the SPM integration (minus the need for the `import`)
- Bonus: Test the Objective-C integration

### Integration Demo

Here's a demo showing the Swift integration using SPM and by manually dragging in the two files noted in the `README`.

https://github.com/doximity/CallWithDoxDialer/assets/3907680/3559c7ff-77a3-417b-a63d-ddae368b2cfa

### CODE REVIEW

**Comment by ronnie3786** [comment]
@dmallory I added you as a reviewer. Can you provide an extra set of eyes on the SPM setup and integration? Let me know if there are any improvements that can be made.

File: Sources/CallWithDoxDialer/CallWithDoxDialer.swift [hunk 1]
```diff
@@ -0,0 +1,117 @@
+import Foundation
+#if canImport(UIKit)
```
Comment (ronnie3786) [inline Sources/CallWithDoxDialer/CallWithDoxDialer.swift]: This wasn't required for integration into another project, but I couldn't get `swift build` to pass validation unless I added this check.

File: Sources/CallWithDoxDialer/CallWithDoxDialer.swift [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+                compatibleWith: nil
+            )
+        else {
+            return UIImage()
+        }
+        return image
+    #else
+        return nil
+    #endif
+    }
```
Comment (ronnie3786) [inline Sources/CallWithDoxDialer/CallWithDoxDialer.swift]: This `#if SWIFT_PACKAGE` was needed when I added the manual integration option. The icon is stored in the Package Bundle (`.module`) but that bundle doesn't exist outside of the Package and would create a compile error when pulling in the files manually.

Manual integration uses the `var iconFromBundle: UIImage?` alternate.

File: .gitignore [hunk 3]
```diff
@@ -17,7 +17,14 @@ profile
...
+
+# OS X
+.DS_Store
+
 ### Carthage ###
 # Carthage - A simple, decentralized dependency manager for Cocoa
 Carthage/Checkouts/
-Carthage/Build/
\ No newline at end of file
+Carthage/Build/
```
Comment (dmallory) [inline .gitignore]: Probably could just delete this one, makes it look like we're using `Carthage`.

File: Package.swift [hunk 4]
```diff
@@ -0,0 +1,30 @@
+// swift-tools-version: 5.9
+// The swift-tools-version declares the minimum version of Swift required to build this package.
+
+import PackageDescription
+
+let package = Package(
+    name: "CallWithDoxDialer",
```
Comment (dmallory) [inline Package.swift]: For Comms to decide, but if you wanted to pick a different/better name for this project and also use it for the package, now would be a good time. The newer Ads SDK is `ios-ads-sdk` and the package is `DoximityAdsSDK` so something like `DoximityCallWithDialer` (if people really like the current name) or `DoximityDialerSDK` (which give more room for future additions) might follow that.

File: Sources/CallWithDoxDialer/CallWithDoxDialer.swift (hunk 1 above)
Comment (dmallory) [reply Sources/CallWithDoxDialer/CallWithDoxDialer.swift]: This feels to me like it's trying to remain valid on other platforms somewhere (MacOS)? but not worth a ton of time to figure out if it's working fine. I saw that it has `iOS` as the platform in the package config. 🤷🏼‍♂️

File: Sources/CallWithDoxDialer/CallWithDoxDialer.swift [hunk 5]
```diff
@@ -0,0 +1,117 @@
+import Foundation
+#if canImport(UIKit)
+import UIKit
+
+/// A struct to handle dialing phone numbers using the Doximity app.
+public struct DoxDialerCaller {
```
Comment (dmallory) [inline Sources/CallWithDoxDialer/CallWithDoxDialer.swift]: Similar to the package name, if you want to follow the newer Ads SDK they made a product decision that we want to spell out `Doximity` as much as possible and get away from `Dox` in any public APIs. The types there are spelled out like https://github.com/doximity/ios-ads-sdk/blob/develop/DoximityAdsSDK/Sources/DoximityAdsPortal.swift if you want to follow it.

**Review by dmallory — APPROVED** [review]
Functionality looks good to me, but worth thinking about naming as that's harder to keep changing in a public API.

Also should make sure this bridging does really work in an Objective-C project. If you want to easily keep that maintained you could include a simple Objective-C example app target as well, but at least that we've tested it initially.

File: Package.swift (hunk 4 above)
Comment (Chandlerdea) [reply Package.swift]: I agree that we should update the name to be more inline with the naming convention of our current SDKs.

File: Sources/CallWithDoxDialer/CallWithDoxDialer.swift [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+            openURL(dialerURL)
+        } else {
+            guard let url = doximityAppStoreURL else { return }
+            openURL(url)
+        }
+    }
+
+    /// Returns the Doximity icon image.
+    public var doximityIcon: UIImage {
+        return iconFromPackage ?? iconFromBundle ?? UIImage()
```
Comment (Chandlerdea) [inline Sources/CallWithDoxDialer/CallWithDoxDialer.swift]: Should we throw an error her and let the consumers handle the case where we can't load the icon from the bundle?

**Review by Chandlerdea — APPROVED** [review]
Looks great, just one question but not a blocker.

**Review by mundue — APPROVED** [review]

File: Package.swift (hunk 4 above)
Comment (ronnie3786) [reply Package.swift]: @dmallory I updated the SDK to be renamed to `DoximityDialerSDK` in this commit: https://github.com/doximity/CallWithDoxDialer/pull/22/commits/a93363a3978cd16c28abdf8c696b5c496c588c38

**Review by Chandlerdea — APPROVED** [review]
👍

**Review by dmallory — APPROVED** [review]

**Pull request merged**